lib/trace.js

```diff
@@ -1,24 +1,114 @@
 "use strict";
 
-const { exists } = require("./util");
+const fs = require("fs");
+const path = require("path");
+const { promisify } = require("util");
+
+const { parse } = require("acorn-node");
+const { getDeps } = require("./extract");
+
+// Helpers
+const resolve = promisify(require("resolve"));
+const readFile = promisify(fs.readFile);
+const uniq = (items) => Array.from(new Set(items)).sort();
+
+// HELPER: Recursively track and trace dependency paths.
+// - `srcPaths`: If provided, **don't** add to resolvedDepPaths
+// - `depPaths`: If provided, **do** add to resolvedDepPaths
+const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, tracedDepPaths }) => {
+  // Start **only** with depPaths.
+  let resolvedDepPaths = Array.from(depPaths);
+
+  // If srcPaths is provided, now **replace** depPaths.
+  // (Allows easier accomodation of `traceFile` vs. `traceFiles`)
+  depPaths = srcPaths || depPaths;
+
+  // TODO(6): Consider parallelizing file traversal
+  // https://github.com/FormidableLabs/trace-deps/issues/6
+  for (const depPath of depPaths) {
+    if (!tracedDepPaths.has(depPath) && (/\.(js|mjs)$/).test(depPath)) {
+      // Mark encountered to avoid future recursion.
+      tracedDepPaths.add(depPath);
+
+      // Recurse.
+      // eslint-disable-next-line no-use-before-define
+      const recursedDepPaths = await traceFile({ srcPath: depPath, ignores, tracedDepPaths });
+
+      // Aggregate.
+      resolvedDepPaths = resolvedDepPaths.concat(recursedDepPaths);
+    }
+  }
+
+  return resolvedDepPaths;
+};
 
 /**
  * Trace and return on-disk locations of all file dependencies from a source file.
  *
- * @param {*}       opts              options object
- * @param {string}  opts.file         source file to trace
- * @returns {Promise<Array<String>>}  list of relative paths to on-disk dependencies
+ * @param {*}                         opts                options object
+ * @param {string}                    opts.srcPath        source file path to trace
+ * @param {Array<string>}             opts.ignores        list of package prefixes to ignore
+ * @param {Set}                       opts.tracedDepPaths tracked dependencies
+ * @returns {Promise<Array<string>>}  list of absolute paths to on-disk dependencies
  */
-const traceFile = async ({ file } = {}) => {
-  if (!file) {
-    throw new Error("Empty file");
-  } else if (!await exists(file)) {
-    throw new Error(`Could not find source file: ${file}`);
+const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } = {}) => {
```

Comment:
Maybe a TODO: to memoize this to prevent circular deps or re-parsing the same file twice?

Reply:
It's not completely obvious, but https://github.com/FormidableLabs/trace-deps/pull/7/files/c5eaa6e16a2f2e5f0a32e60552020a8060eea188#diff-fe45162970b441f082c6c624644a2ab5R87 does mark this file as "traversed" in `tracedDepPaths` and it won't be traversed again. Basically, once only non-encountered files are further inspected...

Reply:
:100: I see it now!

Reply:
I guess that's in the case a shared `tracedDepPaths` is passed around, which includes:

* Any use of `traceFiles`
* Any recursion into `_recurseDeps`
* Multiple calls to `traceFile` with `tracedDepPaths` actually passed manually

This wouldn't cover a narrow case like the following:

```js
// The same file manually traced repeatedly.
const results = await Promise.all(["one.js", "one.js"].map((f) => traceFile({ srcPath: f }));
```

... but I don't think it's that likely an expected use case we would need to support

What do you think? Should I leave as is, add a comment, or a ticket to optimize this last case?

Reply:
Doesn't seem worth addressing that to me!